### PR TITLE
cc-wrapper: improve response file parsing speed

### PR DESCRIPTION
**Pull request by ryantrinkle**
###### Motivation for this change

Although #25205 improved the correctness of response file handling in cc-wrapper, it also devastated performance for large response files, e.g. building Haskell packages with many dependencies.  This commit replaces the bash-based parser with one written in C.  Although bootstrapping is somewhat more complex with this approach, the new code is ~75,000x faster for my test case.

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

CC @orivej

### CODE REVIEW

**Comment by Ericson2314** [comment]
Ah, the merge conflict is because @orivej also pushed a performance improvement to staging. This should still make it quite a bit faster than that too.

**Comment by pbogdan** [comment]
FYI I believe this might be the PR @Ericson2314 was talking about https://github.com/NixOS/nixpkgs/pull/26554

**Comment by Ericson2314** [comment]
A question to decide on is whether we want to keep around the bash version as a fall back, rather than breaking. The error message says "not allowed when bootstrapping", but then again it's actually quite early in the bootstrapping process that this is allowed.

**Comment by orivej** [comment]
If bash parser in `staging` needs to be replaced, I published a simpler C++ one at https://github.com/orivej/expand-compiler-args

**Comment by ryantrinkle** [comment]
@orivej I have no complaint if you'd rather use your code.  My goal was just to solve the issue while keeping the code, dependencies, and bootstrapping logic as simple as possible.  I'm not familiar enough with C++'s toolchain to know whether it would present any difficulties.  I did have to restrain myself from trying to do it in Haskell, though—attoparsec would've made this quite a bit nicer!

BTW, the file I've been testing with is [responsefile.txt](https://github.com/NixOS/nixpkgs/files/1116415/responsefile.txt).

**Comment by orivej** [comment]
I had not considered that C++ toolchain is not available as soon as C toolchain is. I appreciate the work you did to integrate compiled rsp parser into bootstrap process, thank you! Yet `parseResponseFile.c` and `expandResponseParams()` could be much simpler, so I suggest that you replace them with my implementation, which is now in C.

**Comment by 0xABAB** [comment]
I preferred the C++ version, assuming it had the same semantics.

**Comment by orivej** [comment]
If C++ is an option for early Nix bootstrap, I made updated version available at the [cpp branch](https://github.com/orivej/expand-compiler-args/blob/cpp/expand-compiler-args.cpp). The only semantic change was about handling incorrect .rsp files as gcc, e.g. `good arg and 'unclosed with trailing\` is now parsed as `normal arg and unclosed arg with trailing`.

Current timings of parsing [responsefile.txt](https://github.com/NixOS/nixpkgs/files/1116415/responsefile.txt) with C, `staging`, and `master` versions of `expandResponseParams` are: 0m0.013s, 0m2.110s, and 3m25.268s. (Actual C parser completes under 2ms, the rest must be Bash overhead.)

**Comment by orivej** [comment]
Alternatively, bash version could be optimized somewhat further:
https://github.com/orivej/expand-compiler-args/commit/d77495a8b7d295b0f4f10d7e6d9c225f0e8640df reduces `responsefile.txt` parsing time from 2s to 1.2s, `ghc_7.rsp` from 1s to 0.6s.

**Comment by 0xABAB** [comment]
Can we get anything which improves the situation *now* as opposed to what might be nice to have some later date in the future? The initial 75,000 factor of improvement seems to be interesting enough to include regardless of whether it could be made faster in another PR. 

At least that's the process I would like to see.

**Comment by orivej** [comment]
> Can we get anything which improves the situation now as opposed to what might be nice to have some later date in the future?
Yes, merge nixpkgs `staging` branch into `master`.
> The initial 75,000 factor of improvement seems to be interesting enough to include regardless of whether it could be made faster in another PR.
This PR is for the `staging` branch, as it should be because it causes everything to be rebuilt. Merging it will reduce parsing time of the provided sample from 2s to 0.02s.

**Comment by 0xABAB** [comment]
OK, so we are now just waiting for someone to push the merge button then, unless there are more comments from others.

**Comment by orivej** [comment]
No, merging this PR has nothing to do with merging `staging` into `master`. This PR is about speeding up `staging`.

**Comment by 0xABAB** [comment]
@orivej I did not say there was, since the top says `ryantrinkle  wants to merge 1 commit into NixOS:staging from obsidiansystems:response-file-parsing-speed`. I am well aware as to what the role is of `staging`.

**Comment by orivej** [comment]
I'd wait for @ryantrinkle on https://github.com/obsidiansystems/nixpkgs/pull/3, and for people here to say if they want this additional speedup or if they do not want this additional complexity. For my purposes current `staging` is good enough.

**Review by peti — APPROVED** [review]
I am in favor of merging this change. I don't fully understand the implications of adding a C program into the cc wrapper, but intuitively I'd assume that this is no problem. It would be nice to hear what @vcunat thinks about that issue.

File: pkgs/build-support/cc-wrapper/default.nix
Comment (orivej) [inline pkgs/build-support/cc-wrapper/default.nix]: What do you mean? Why not `"$CC" -o "$out" "$src"`?

Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: We got a runtime dependency on the source file, probably because `__FILE__` in the assert expansion.

**Comment by Ericson2314** [comment]
I think C++ should be fine.  Also, if I recall correctly, there is some `langCC` attribute or similar on compilers that can be used to tell whether the compiler is supposed to support it.

File: pkgs/build-support/cc-wrapper/default.nix
Comment (orivej) [inline pkgs/build-support/cc-wrapper/default.nix]: That's plausible, but why is it undesirable?

**Comment by orivej** [comment]
This PR is ready for merge.

**Comment by domenkozar** [comment]
Let's wait on approval from @Ericson2314 and @ryantrinkle and merge :)

**Comment by ryantrinkle** [comment]
@domenkozar This is good to go.

File: pkgs/build-support/cc-wrapper/expand-response-params.c
Comment (nh2) [inline pkgs/build-support/cc-wrapper/expand-response-params.c]: What's this custom string type (above) + append for? Why can't this use `std::string`?

Comment (nh2) [inline pkgs/build-support/cc-wrapper/expand-response-params.c]: Ah, I got confused that there's a proposed C++ implementation, but this one is not that, this is a C implementation. So nevermind.

**Comment by nh2** [comment]
I found that even after this PR merged (nixpkgs commit aad2a2132c4 for me), `ld-wrapper` invocations can still take 10 seconds of CPU time in bash for each executable for large Haskell projects.

**Comment by nh2** [comment]
> invocations can still take 10 seconds of CPU time in bash for each executable for large Haskell projects

Fixed in #27657.

**Pull request merged**
